File: bridge/cpp/examples/jacobi_stencil.cpp
#include <iostream>
#include "bh/bh.hpp"

using namespace std;
using namespace bh;

template <typename T>
multi_array<T>& iterate(int w, int h, int i)
{
    multi_array<T> grid(h+2,w+2), center, north, south, east, west;

    grid = 0.0;
    grid[_(0,h,1)][ 0] = -273.15;
    grid[_(0,h,1)][-1] = -273.15;
    grid[-1][_(0,w,1)] = -273.15;
    grid[ 0][_(0,w,1)] =    40.0;

    center  = grid[_(1,-1,1)][_(1,-1,1)];
    north   = grid[_(0,-2,1)][_(1,-1,1)];
    south   = grid[_(2, h,1)][_(1,-1,1)];
    east    = grid[_(1,-1,1)][_(2, w,1)];
    west    = grid[_(1,-1,1)][_(0,-2,1)];

    for(int k=0; k<i; k++) {
        center = (T)0.2*(center+north+east+west+south);
    }

    return grid;
}

int main(int argc, char* argv[])
{
    int height, width, iterations;
    if (argc < 4) {
        cout << "Error: Not enough argumnts, call like: " << endl;
        cout << argv[0] << " 600 400 2" << endl;
        cout << "For a 600x400 grid and 2 iterations." << endl;
        return 0;
    }
    width       = atoi(argv[1]);
    height      = atoi(argv[2]);
    iterations  = atoi(argv[3]);

    cout << "[Jacobi Stencil(" \
            "width=" << width << \
            ",height=" << height << \
            ",iter=" << iterations << \
            ")]" << endl;
    iterate<double>(width, height, iterations);
    cout << "done!" << endl;

    return 0;
}

File: bridge/bhpy/examples/jacobi_stencil.py
"""
Jacobi Stencil
--------------

So what does this code example illustrate?
"""
import bohrium as np

def freezetrap(height, width, dtype=np.float32, bohrium=True):
    grid = np.zeros((height+2,width+2), dtype=dtype, bohrium=bohrium)
    grid[:,0]  = dtype(-273.15)
    grid[:,-1] = dtype(-273.15)
    grid[-1,:] = dtype(-273.15)
    grid[0,:]  = dtype(40.0)
    return grid

def solve(grid, epsilon=0.005, max_iterations=None, visualize=False):
    center = grid[1:-1, 1:-1]
    north  = grid[0:-2, 1:-1]
    east   = grid[1:-1, 2:  ]
    west   = grid[1:-1, 0:-2]
    south  = grid[2:  , 1:-1]
    delta = epsilon + 1
    iteration = 0
    while delta > epsilon:
        iteration += 1
        work = 0.2*(center+north+east+west+south)
        delta = np.add.reduce(np.add.reduce(np.absolute(work-center)))
        center[:] = work
        if max_iterations != None and max_iterations <= iteration:
            break
        if visualize:
            np.visualize(grid, "2d", 0, 0.0, 5.5)
    return grid

def iterate(grid, iterations, visualize=False):
    center = grid[1:-1, 1:-1]
    north  = grid[0:-2, 1:-1]
    east   = grid[1:-1, 2:  ]
    west   = grid[1:-1, 0:-2]
    south  = grid[2:  , 1:-1]
    for i in xrange(iterations):
        center[:] = 0.2*(center+north+east+west+south)
        if visualize:
            np.visualize(grid, "2d", 0, 0.0, 5.5)
    return grid

